x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/TransportMountSearchableSnapshotAction.java

```diff
@@ -0,0 +1,158 @@
+    }
+
+    @Override
+    protected String executor() {
+        // Using the generic instead of the snapshot threadpool here as the snapshot threadpool might be blocked on long running tasks
+        // which would block the request from getting an error response because of the ongoing task
+        return ThreadPool.Names.GENERIC;
+    }
+
+    @Override
+    protected RestoreSnapshotResponse read(StreamInput in) throws IOException {
+        return new RestoreSnapshotResponse(in);
+    }
+
+    @Override
+    protected ClusterBlockException checkBlock(MountSearchableSnapshotRequest request, ClusterState state) {
+        // Restoring a snapshot might change the global state and create/change an index,
+        // so we need to check for METADATA_WRITE and WRITE blocks
+        ClusterBlockException blockException = state.blocks().globalBlockedException(ClusterBlockLevel.METADATA_WRITE);
+        if (blockException != null) {
+            return blockException;
+        }
+        return state.blocks().globalBlockedException(ClusterBlockLevel.WRITE);
+
+    }
+
+    /**
+     * Return the index settings required to make a snapshot searchable
+     */
+    private static Settings getIndexSettings(String repoName, SnapshotId snapshotId, IndexId indexId) {
+        return Settings.builder()
+            .put(SearchableSnapshotRepository.SNAPSHOT_REPOSITORY_SETTING.getKey(), repoName)
+            .put(SearchableSnapshotRepository.SNAPSHOT_SNAPSHOT_NAME_SETTING.getKey(), snapshotId.getName())
+            .put(SearchableSnapshotRepository.SNAPSHOT_SNAPSHOT_ID_SETTING.getKey(), snapshotId.getUUID())
+            .put(SearchableSnapshotRepository.SNAPSHOT_INDEX_ID_SETTING.getKey(), indexId.getId())
+            .put(INDEX_STORE_TYPE_SETTING.getKey(), SearchableSnapshotRepository.SNAPSHOT_DIRECTORY_FACTORY_KEY)
+            .put(IndexMetaData.SETTING_BLOCKS_WRITE, true)
+            .build();
+    }
+
+    @Override
+    protected void masterOperation(Task task, final MountSearchableSnapshotRequest request, final ClusterState state,
+                                   final ActionListener<RestoreSnapshotResponse> listener) {
+        final String repoName = request.repository();
+        final String snapName = request.snapshot();
+        final String indexName = request.snapshotIndex();
+
+        // Retrieve IndexId and SnapshotId instances, which are then used to create a new restore
+        // request, which is then sent on to the actual snapshot restore mechanism
+        final Repository repository = repositoriesService.repository(repoName);
+        final StepListener<RepositoryData> repositoryDataListener = new StepListener<>();
+        repository.getRepositoryData(repositoryDataListener);
+        repositoryDataListener.whenComplete(repoData -> {
+            final Map<String, IndexId> indexIds = repoData.getIndices();
+
+            if (indexIds.containsKey(indexName) == false && indexIds.get(indexName) != null) {
```

Comment:
I'm wondering if we need these verification - maybe we could just let the Restore action do it for us? (and configure appropriate indices options like ignore_unavailable: false)

Reply:
If we don't have these then we can't fetch the `IndexId`, so essentially we'd be sending a half-broken restore (because it would be missing crucial index settings required for searchable snapshots to work) in order to push the error elsewhere, I'd rather do the verification as soon as possible so we don't pass around half-broken requests (and do one fewer internal transport requests)

Reply:
Yes, sorry for the noise. I've been too quick at commenting here.